Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp

```diff
@@ -0,0 +1,226 @@
+    // then the given effect (e.g. a gradient or a filter) will be ignored.
+    FloatRect objectBoundingBox = renderer.objectBoundingBox();
+    if (m_attributes->patternUnits() == SVGUnitTypes::SVG_UNIT_TYPE_OBJECTBOUNDINGBOX && objectBoundingBox.isEmpty())
+        return nullptr;
+
+    // If we couldn't determine the pattern content element root, stop here.
+    if (!m_attributes->patternContentElement())
+        return nullptr;
+
+    // An empty viewBox disables rendering.
+    if (m_attributes->hasViewBox() && m_attributes->viewBox().isEmpty())
+        return nullptr;
+
+    // Compute all necessary transformations to build the tile image & the pattern.
+    FloatRect tileBoundaries;
+    AffineTransform tileImageTransform;
+    if (!buildTileImageTransform(renderer, *m_attributes, patternElement(), tileBoundaries, tileImageTransform))
+        return nullptr;
+
+    // Ignore 2D rotation, as it doesn't affect the size of the tile.
+    auto absoluteTransformIgnoringRotation = context.getCTM(GraphicsContext::DefinitelyIncludeDeviceScale);
+    SVGRenderingContext::clear2DRotation(absoluteTransformIgnoringRotation);
+    FloatRect absoluteTileBoundaries = absoluteTransformIgnoringRotation.mapRect(tileBoundaries);
+
+    // Scale the tile size to match the scale level of the patternTransform.
+    absoluteTileBoundaries.scale(static_cast<float>(m_attributes->patternTransform().xScale()), static_cast<float>(m_attributes->patternTransform().yScale()));
+
+    // Build tile image.
+    auto tileImage = createTileImage(*m_attributes, tileBoundaries, absoluteTileBoundaries, tileImageTransform);
+    if (!tileImage)
+        return nullptr;
+
+    auto tileImageSize = tileImage->logicalSize();
+
+    auto copiedImage = ImageBuffer::sinkIntoNativeImage(WTFMove(tileImage));
+    if (!copiedImage)
+        return nullptr;
+
+    // Compute pattern space transformation.
+    AffineTransform patternSpaceTransform;
+    patternSpaceTransform.translate(tileBoundaries.location());
+    patternSpaceTransform.scale(tileBoundaries.size() / tileImageSize);
+
+    AffineTransform patternTransform = m_attributes->patternTransform();
+    if (!patternTransform.isIdentity())
+        patternSpaceTransform = patternTransform * patternSpaceTransform;
+
+    // Build pattern.
+    return Pattern::create({ copiedImage.releaseNonNull() }, { true, true, patternSpaceTransform });
+}
+
+bool RenderSVGResourcePattern::buildPatternIfNeeded(GraphicsContext& context, const RenderLayerModelObject& targetRenderer)
+{
+    collectPatternAttributesIfNeeded();
+
+    m_pattern = buildPattern(context, targetRenderer);
+
+    if (!m_pattern)
+        return false;
+
+    return true;
+}
+
+bool RenderSVGResourcePattern::prepareFillOperation(GraphicsContext& context, const RenderLayerModelObject& targetRenderer, const RenderStyle& style)
+{
+    if (SVGHitTestCycleDetectionScope::isVisiting(*this))
```

Comment:
I think this is not needed if you resolve cycles during collectPatternAttr - or am I mistaken?

Reply:
I believe putting all logic in collectPatternAttr on DOM side will not catch some cases from custom/pattern-content-inheritance-cycle.svg like:
<!-- a => b => a -->
        <pattern id="a" x="0" y="0" width=".25" height=".25">
            <rect fill="lime" width="100%" height="100%"/>
            <rect fill="url(#b)" width="100%" height="100%"/>
        </pattern>
        <pattern id="b" xlink:href="#a"/>
At least when I tried it crashed :)

Reply:
Ok fair enough

Reply:
I just wanted to suggest to unify the code with gradient, to avoid the URI resolving in collectGradientAttributes, but...

I think, only patterns have the ability to have children that can reference other resources - gradients only have stop children.

Reply:
That would mean everything is fine, we only need to check the coverage for cycles and gradients is sufficient.